mldsa/reduce.h

```diff
@@ -6,16 +6,27 @@
 #define REDUCE_H
 
 #include <stdint.h>
+#include "cbmc.h"
 #include "params.h"
 
 #define MONT -4186625  // 2^32 % Q
 #define QINV 58728449  // q^(-1) mod 2^32
+#define REDUCE_DOMAIN_MAX (INT32_MAX - (1 << 22))
+#define REDUCE_RANGE_MAX 6283009
 
 #define montgomery_reduce DILITHIUM_NAMESPACE(montgomery_reduce)
 int32_t montgomery_reduce(int64_t a);
 
 #define reduce32 DILITHIUM_NAMESPACE(reduce32)
-int32_t reduce32(int32_t a);
+int32_t reduce32(int32_t a)
+__contract__(
+  requires(a <= REDUCE_DOMAIN_MAX)
+  ensures(return_value >= -REDUCE_RANGE_MAX)
+  ensures(return_value <   REDUCE_RANGE_MAX)
+  ensures((return_value == (a % Q)) ||
```

Comment:
Is that really what you want to say, or can you instead use a combination of (a) (return_value - a) % Q == 0 and (b) a bounds condition on `return_value`?

You probably only need the bounds condition elsewhere, because we don't do functional correctness proofs much in CBMC otherwise.

Reply:
Thank you, added in [c3f0fbb](https://github.com/pq-code-package/mldsa-native/pull/60/commits/c3f0fbb8886fe7463f678bacf4d9537e485eb352).